### PR TITLE
fix(core): fix the bug of incorrect result set of count function (#736)

### LINKED ISSUE #736: bug: Insert a space，SELECT COUNT(DISTINCT col2) FROM t1;Error query result
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/stoneatom/stonedb/blob/stonedb-5.7-dev/CONTRIBUTING.md).

### Please confirm if bug report does NOT exists already ?

- [X] I confirm there is no existing issue for this

### Describe the problem

```
CREATE TABLE t1(col1 INT, col2 CHAR(5))ENGINE=tianmu;
INSERT INTO t1 VALUES(NULL,''),(1,'a'),(1,'b'),(1,'c'),(2,'dd'),(3,'eee');
SELECT COUNT(DISTINCT col2) FROM t1;

##The error results are as follows
+----------------------+
| COUNT(DISTINCT col2) |
+----------------------+
|                    1 |
+----------------------+
```

### Expected behavior

```
##The following is：innodb test result,mariadb-Columnstore test result is 5

mysql> SELECT COUNT(DISTINCT col2) FROM t1;
+----------------------+
| COUNT(DISTINCT col2) |
+----------------------+
|                    6 
+----------------------+
```

### How To Reproduce

```
```
CREATE TABLE t1(col1 INT, col2 CHAR(5))ENGINE=tianmu;
INSERT INTO t1 VALUES(NULL,''),(1,'a'),(1,'b'),(1,'c'),(2,'dd'),(3,'eee');
SELECT COUNT(DISTINCT col2) FROM t1;
```

### Environment

```
Ver 14.14 Distrib 5.7.36-StoneDB
```

### Are you interested in submitting a PR to solve the problem?

- [ ] Yes, I will!

### PR DESCRIPTION
**Pull request by lylth**
Replace the function FindCurrentRowByVMTable that gets the current row to FindCurrentRow.

## Summary about this PR

Issue Number: close #736 
fix(core): fix the bug of incorrect result set of count function (#736)
         Replace the function FindCurrentRowByVMTable that gets the current row to FindCurrentRow.

## Tests Check List

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [x] Bug Fix
- [ ] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

**Comment by RingsC** [comment]
@lylth  Ref : https://github.com/stoneatom/stonedb/pull/472. Changes in this PR.

**Review by isredstar — APPROVED** [review]
LGTM

File: storage/tianmu/core/group_distinct_table.cpp [hunk 1]
```diff
@@ -55,11 +55,13 @@ GroupDistinctTable::~GroupDistinctTable() {
 void GroupDistinctTable::InitializeVC(int64_t max_no_groups, vcolumn::VirtualColumn *vc, int64_t max_no_rows,
                                       int64_t max_bytes, bool decodable) {
   DEBUG_ASSERT(!initialized);
-  if (max_bytes > 0) max_total_size = max_bytes;
+  if (max_bytes > 0)
+    max_total_size = max_bytes;
   if (max_bytes > 2000000000)  // possible for large aggregation settings, but
```
Comment (RingsC) [inline storage/tianmu/core/group_distinct_table.cpp]: using : 2 * 1024 * 1024 *1024, instead of 2000000000.

File: storage/tianmu/core/group_distinct_table.cpp [hunk 2]
```diff
@@ -132,21 +134,23 @@ void GroupDistinctTable::InitializeBuffers(int64_t max_no_rows)  // max_no_rows
   // calculate vertical size (not dividable by 17)
   if (no_rows < 67)  // too less groups => high collision probability; 67 is prime.
     no_rows = 67;
-  if (no_rows % 17 == 0) no_rows++;
+  if (no_rows % 17 == 0)
+    no_rows++;
   rows_limit = int64_t(no_rows * 0.9);  // rows_limit is used to determine whether the table is full
 
   t = (unsigned char *)alloc(total_width * no_rows, mm::BLOCK_TYPE::BLOCK_TEMPORARY);
```
Comment (RingsC) [inline storage/tianmu/core/group_distinct_table.cpp]: using c++ style type cast, and replace all old style casting with c++ style in this file.

File: storage/tianmu/core/group_distinct_table.cpp [hunk 3]
```diff
@@ -132,21 +134,23 @@ void GroupDistinctTable::InitializeBuffers(int64_t max_no_rows)  // max_no_rows
...
   //	t = new BlockedRowMemStorage(total_width, &mem_mngr, no_rows);
   input_buffer = (unsigned char *)(new int[total_width / 4 + 1]);  // ensure proper memory alignment
```
Comment (RingsC) [inline storage/tianmu/core/group_distinct_table.cpp]: c++ style type cast.

File: mysql-test/suite/tianmu/t/issue736.test [hunk 4]
```diff
@@ -0,0 +1,7 @@
+--source include/have_tianmu.inc
+use test;
+CREATE TABLE t1(col1 INT, col2 CHAR(5))ENGINE=tianmu;
+INSERT INTO t1 VALUES(NULL,''),(1,'a'),(1,'b'),(1,'c'),(2,'dd'),(3,'eee');
```
Comment (RingsC) [inline mysql-test/suite/tianmu/t/issue736.test:4]: Add some tests for unicode charset.

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/group_distinct_table.cpp [hunk 5]
```diff
@@ -132,21 +134,23 @@ void GroupDistinctTable::InitializeBuffers(int64_t max_no_rows)  // max_no_rows
   // calculate vertical size (not dividable by 17)
   if (no_rows < 67)  // too less groups => high collision probability; 67 is prime.
     no_rows = 67;
-  if (no_rows % 17 == 0) no_rows++;
+  if (no_rows % 17 == 0)
+    no_rows++;
   rows_limit = int64_t(no_rows * 0.9);  // rows_limit is used to determine whether the table is full
 
-  t = (unsigned char *)alloc(total_width * no_rows, mm::BLOCK_TYPE::BLOCK_TEMPORARY);
+  t = reinterpret_cast<unsigned char *>(alloc(total_width * no_rows, mm::BLOCK_TYPE::BLOCK_TEMPORARY));
```
Comment (RingsC) [inline storage/tianmu/core/group_distinct_table.cpp]: Use static_cast, not reinterpret_cast, 
Ref: https://en.cppreference.com/w/cpp/language/reinterpret_cast
        https://en.cppreference.com/w/cpp/language/static_cast

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/core/group_distinct_table.cpp [hunk 6]
```diff
@@ -132,21 +134,23 @@ void GroupDistinctTable::InitializeBuffers(int64_t max_no_rows)  // max_no_rows
   // calculate vertical size (not dividable by 17)
   if (no_rows < 67)  // too less groups => high collision probability; 67 is prime.
     no_rows = 67;
-  if (no_rows % 17 == 0) no_rows++;
+  if (no_rows % 17 == 0)
+    no_rows++;
   rows_limit = int64_t(no_rows * 0.9);  // rows_limit is used to determine whether the table is full
 
-  t = (unsigned char *)alloc(total_width * no_rows, mm::BLOCK_TYPE::BLOCK_TEMPORARY);
+  t = static_cast<unsigned char *>(alloc(total_width * no_rows, mm::BLOCK_TYPE::BLOCK_TEMPORARY));
```
Comment (RingsC) [inline storage/tianmu/core/group_distinct_table.cpp:141]: to set  initial memory, set to 0x0.

File: storage/tianmu/core/group_distinct_table.cpp [hunk 7]
```diff
@@ -132,21 +134,23 @@ void GroupDistinctTable::InitializeBuffers(int64_t max_no_rows)  // max_no_rows
...
   //	t = new BlockedRowMemStorage(total_width, &mem_mngr, no_rows);
-  input_buffer = (unsigned char *)(new int[total_width / 4 + 1]);  // ensure proper memory alignment
-  rc_control_.lock(m_conn->GetThreadID()) << "GroupDistinctTable initialized as Hash(" << no_rows << "), " << group_bytes
-                                        << "+" << value_bytes << " bytes." << system::unlock;
+  input_buffer = reinterpret_cast<unsigned char *>(new int[total_width / 4 + 1]);  // ensure proper memory alignment
```
Comment (RingsC) [inline storage/tianmu/core/group_distinct_table.cpp:144]: Same thing.

**Review by RingsC — CHANGES_REQUESTED** [review]

**Review by RingsC — APPROVED** [review]
